### PR TITLE
Improved Polyface.ts

### PR DESCRIPTION
**Pull request by saeeedtorabi**
- Improved formatting and documentations of Polyface.ts and other related classes.
- Added md doc for Polyface.
- Fixed few minor bugs.
- Added few tests.

### CODE REVIEW

File: core/geometry/src/polyface/Polyface.ts [hunk 1]
```diff
@@ -316,126 +355,145 @@ export class IndexedPolyface extends Polyface {
...
   public addColorIndex(index: number): void {
     if (!this.data.colorIndex)
       this.data.colorIndex = [];
     this.data.colorIndex.push(index);
   }
-
-  /** clean up the open facet.  return the returnValue (so caller can easily return cleanupOpenFacet("message")) */
+  /**
+   * Clean up the open facet.
+   * @deprecated in 4.x. Its single caller does not need to call this API.
```
Comment (saeeedtorabi) [inline core/geometry/src/polyface/Polyface.ts:474]: Lint forced me to provide a reason for deprecate. Not sure if this is best description of the reason!

File: core/geometry/src/polyface/Polyface.ts [hunk 2]
```diff
@@ -30,94 +30,143 @@ import { PolyfaceData } from "./PolyfaceData";
...
+   * Check validity of indices into a data array.
+   * * It is valid to have both indices and data undefined.
+   * * It is NOT valid for just one to be defined.
+   * * Index values at indices[indexPositionA <= i < indexPositionB] must be valid indices to the data array.
+   * @param indices array of indices.
+   * @param indexPositionA first index to test.
+   * @param indexPositionB one past final index to test.
+   * @param data data array.
+   * @param dataLength length of data array.
+   */
```
Comment (saeeedtorabi) [inline core/geometry/src/polyface/Polyface.ts:73]: Failure of RotationalSweeps test might be related to changes in this function too. I will investigate.

File: core/geometry/src/polyface/Polyface.ts [hunk 3]
```diff
@@ -316,126 +355,145 @@ export class IndexedPolyface extends Polyface {
...
+  /**
+   * Add (a clone of) a normal vector to the normal array.
+   * @param normal the normal vector.
    * @param priorIndexA first index to check for possible duplicate value.
    * @param priorIndexB second index to check for possible duplicate value.
-   * @returns 0-based index of the added or reused normal.
+   * @returns zero-based index of the added or duplicate normal.
    */
   public addNormal(normal: Vector3d, priorIndexA?: number, priorIndexB?: number): number {
+    // check if `normal` is duplicate of `dataNormal` at index `i`
```
Comment (saeeedtorabi) [inline core/geometry/src/polyface/Polyface.ts:406]: I can even remove the first parameter and change the lambda function to
``` 
    const normalIsDuplicate = (i: number) => {
      const distance = this.data.normal!.distanceIndexToPoint(i, normal);
      return distance !== undefined && Geometry.isSmallMetricDistance(distance);
    };
 ```
 and call it like ```normalIsDuplicate(priorIndexA)```

File: core/geometry/src/polyface/Polyface.ts [hunk 4]
```diff
@@ -444,140 +502,163 @@ export class IndexedPolyface extends Polyface {
...
     return handler.handleIndexedPolyface(this);
   }
 }
 
 /**
  * A PolyfaceVisitor manages data while walking through facets.
- *
  * * The polyface visitor holds data for one facet at a time.
- * * The caller can request the position in the addressed facets as a "readIndex."
- * * The readIndex value (as a number) is not promised to be sequential. (I.e. it might be a simple facet count or might be
```
Comment (saeeedtorabi) [inline core/geometry/src/polyface/Polyface.ts:548]: copied the text from Earling's PR

**Review by dassaf4 — APPROVED** [review]

File: core/geometry/src/polyface/Polyface.ts (hunk 2 above)
Comment (saeeedtorabi) [reply core/geometry/src/polyface/Polyface.ts:73]: I will investigate the failure. Not sure if it relates to the change. If so, I can restore it.

**Pull request merged**